File: Algorithms/two_number_sum/two_number_sum.cpp
//
// Created by edydee on 23/06/21.
//

#include "two_number_sum.h"
#include <unordered_map>

vector<int> two_number_sum::two_number_sum_ind(vector<int> &array, int target) {
    unordered_map<int, int> hash_map;
    for(int i = 0; i < array.size(); i++) {
        int num = array[i];
        int other_half = target - num;
        if(hash_map.find(other_half) != hash_map.end()) {
            return { hash_map[other_half], i };
        }
        hash_map[num] = i;
    }
    return {};
}

File: DSAInJava/src/com/dsainjava/TwoNumberSum/TwoNumberSum.java
package com.dsainjava.TwoNumberSum;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class TwoNumberSum {
    public static ArrayList<Integer> twoNumberSum(ArrayList<Integer> nums, int target) {
        ArrayList<Integer> twoNums = new ArrayList<>();
        HashMap<Integer, Boolean> hashMap = new HashMap<>();
        for (int current : nums) {
            int otherHalf = target - current;
            if (hashMap.containsKey(otherHalf)) {
                twoNums.add(current);
                twoNums.add(otherHalf);
                return twoNums;
            }
            hashMap.put(current, true);
        }

        return twoNums;
    }

    public static void main(String[] args) {
        ArrayList<Integer> nums = new ArrayList<>(List.of(1, 4, 7, -1, 0, -2, 2));
        int target = 11;
        ArrayList<Integer> twoNums = twoNumberSum(nums, target);
        System.out.println(twoNums);
    }
}
